src/util/arrayView.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+
+import {type PluckFn} from "./sortBy";
+import sortBy from "./sortBy";
+
+export class ArrayView<T> {
+  _array: Array<T>;
+  _filteredArray: Array<T>;
+  _rowsPerPage: number;
+  _pages: number;
+
+  constructor(data: $ReadOnlyArray<T>) {
+    this._array = JSON.parse(JSON.stringify(data)); //Deep Copy
+    this._filteredArray = this._array;
+    this._rowsPerPage = this._array.length;
+    this._pages = 1;
+  }
+
+  /**
+    pluckArg: a function of an object that returns the element by which to sort
+    order: "ascending" or "descending"
+   */
+  sortBy(pluckArg: PluckFn<T>, order: "ascending" | "descending") {
+    this._array = sortBy(this._array, pluckArg);
+    this._filteredArray = sortBy(this._filteredArray, pluckArg);
+    if (order === "descending") {
+      this._array.reverse();
+      this._filteredArray.reverse();
```

Comment:
Hmm, sort is `O(n * log n)`, and this function is likely to be in the hot path for re-rendering the frontend (i.e. user changes the sort order in the UI). This function does two sorts, which is unnecessary. In particular, we never need to sort the full array since we don't actually return directly from it, we just need to sort the filtered array. It's easy to imagine cases where the filtered array is dramatically smaller than the full array (e.g. filtered down to just contributions in October 2020, or filtered down to just GitHub pull requests), so this could be a real change in performance.

How about we cache the sortBy arguments, so that way when the filter changes, we can compute a new filteredArray and then re sort it?

Reply:
Good idea, I'll add a cache.